Stored objects update instructions

irbekrm opened the pull request:
This PR adds a section to the release notes about what users may need to do following the API deprecation.

In practice this would affect users who have a `cert-manager` installation that has been previously deployed at `cert-manager` version `v0.16` because the stored versions of custom resources at that time were `v1alpha2`, `v1beta` etc. 

These steps need to be done before upgrading to `cert-manager` `v1.6`. I thought it might be useful to have them ready with `v1.4` release notes as the deprecation warnings are added in `v1.4`, but perhaps this is too early?

This PR is against release-next branch as it adds v1.4 specific documentation. Once v1.4 is released, we should also merge this into master.

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
```diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+
+## Updating cert-manager CRDs and stored versions of cert-manager custom resources
+
+We have deprecated `cert-manager.io/v1alpha2`, `cert-manager.io/v1alpha3`, `cert-manager.io/v1beta1`, `acme.cert-manager.io/v1alpha2`, `acme.cert-manager.io/v1alpha3`, `acme.cert-manager.io/v1beta1` APIs.
```
suggestion: would this be a little more readable as a list?

e.g.

```suggestion
We've deprecated the following APIs: 

- `cert-manager.io/v1alpha2`
- `cert-manager.io/v1alpha3`
- `cert-manager.io/v1beta1`
- `acme.cert-manager.io/v1alpha2`
- `acme.cert-manager.io/v1alpha3`
- `acme.cert-manager.io/v1beta1`

```

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
```diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+
+We have deprecated `cert-manager.io/v1alpha2`, `cert-manager.io/v1alpha3`, `cert-manager.io/v1beta1`, `acme.cert-manager.io/v1alpha2`, `acme.cert-manager.io/v1alpha3`, `acme.cert-manager.io/v1beta1` APIs.
+These APIs will be removed in `cert-manager` `v1.6`.
+If you have a `cert-manager` installation that is using or has previously used these APIs you may need to update `cert-manager` custom resources and CRDs. This needs to be done before upgrading to `cert-manager` `v1.6`. (See Kubernetes docs for more detailed explanation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existi ...
```
suggestion: minor wording change
nitpick: we refer to "`cert-manager`" here, but we frequently use "cert-manager" in the documentation, e.g. in https://cert-manager.io/docs/release-notes/release-notes-1.3/ - should we follow the existing style? (this applies elsewhere in the commit)

```suggestion
If you have a cert-manager installation that is using or has previously used these deprecated APIs you may need to update your cert-manager resources and CRDs. This needs to be done before upgrading to cert-manager `v1.6`. (See the [Kubernetes docs](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version) for a more detailed explanation)
```

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
```diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+
+Steps:
+
+1. Make sure your `cert-manager` deployment is `v1` or later.
```
suggestion: this is referring to the installed version of cert-manager, right? if so, would it be good to give the exact version number as it appears to people, to make it exactly clear which version they need?

```suggestion
1. Make sure your installed `cert-manager` deployment is `v1.0.0` or later.
```

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
```diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+Steps:
+
+1. Make sure your `cert-manager` deployment is `v1` or later.
+2. Make sure any version-controlled `cert-manager` custom resource config files updated to use `cert-manager.io/v1` API and re-applied. This should update stored versions for the given resources.
```
nitpick: missing word

```suggestion
2. Make sure any version-controlled `cert-manager` custom resource config files are updated to use the `cert-manager.io/v1` API and re-applied. This should update stored versions for the given resources.
```

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
```diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+1. Make sure your `cert-manager` deployment is `v1` or later.
+2. Make sure any version-controlled `cert-manager` custom resource config files updated to use `cert-manager.io/v1` API and re-applied. This should update stored versions for the given resources.
+
+Further please follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):
+
+1. If there are other resources (`ingress-shim` `Certificate`s, `CertificateRequest`s etc that aren't stored in Git and may have been created using the deprecated APIs, you can run  `kubectl get <resource_name> -oyaml | kubectl apply -f -` to force the resources to be stored in `etcd` at the current storage version.
+2. To remove legacy API versions from `cert-manager` CRDs, you can run the following `curl` commands:
```
suggestion: a slight rewording; I used `\` only to escape the code block.

(I also wouldn't personally expect the documentation to use "please" here, but that's a nitpick)

```suggestion
After that, follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):

1. If there are other resources (e.g. `Certificate`s created by ingress-shim, `CertificateRequest`s, etc.) that aren't stored in Git and may have been created using the deprecated APIs, run
   \```bash
   kubectl get <resource_name> -oyaml | kubectl apply -f -
   \```
   to force the resources to be stored in `etcd` at the current storage version.

2. To remove legacy API versions from `cert-manager` CRDs run the following `curl` commands:
```

SgtCoDFish commented on content/en/docs/release-notes/release-notes-1.4.md:
````diff
@@ -21,4 +21,35 @@ notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
+
+1. If there are other resources (`ingress-shim` `Certificate`s, `CertificateRequest`s etc that aren't stored in Git and may have been created using the deprecated APIs, you can run  `kubectl get <resource_name> -oyaml | kubectl apply -f -` to force the resources to be stored in `etcd` at the current storage version.
+2. To remove legacy API versions from `cert-manager` CRDs, you can run the following `curl` commands:
+```
````
nitpick: I think it's always best to specify a type for code blocks, e.g.

```suggestion
```bash
```

SgtCoDFish reviewed (changes requested):
A few little things from me.

As a bigger thing: should the "steps" section maybe be moved into its own file and have `release-notes-1.4.md` link to that file? We're going to be linking to these instructions a lot, I imagine, over the coming year, including after the release of 1.5 - it doesn't feel like they really fit in the 1.4 release notes.

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,48 @@
+---
+title: "Remove deprecated API resources"
+linkTitle: "Remove deprecated API resources"
```
nitpick: we currently use title casing everywhere for the sidebar. Also, does this sound better with "removing" instead of "remove", to indicate it's a guide?

```suggestion
title: "Removing Deprecated API Resources"
linkTitle: "Removing Deprecated API Resources"
```

irbekrm replied:
Ahh, yes the title case :)

SgtCoDFish replied:
:see_no_evil: I'm sorry :see_no_evil:

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,48 @@
+- `acme.cert-manager.io/v1alpha2`
+- `acme.cert-manager.io/v1alpha3`
+- `acme.cert-manager.io/v1beta1`
+These APIs will be removed in "cert-manager" `v1.6.0`.
```
suggestion: sorry, I should've been clearer: I meant we could replace `cert-manager` with cert-manager on its own. Like this:

```suggestion
These APIs will be removed in cert-manager `v1.6.0`.
```

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
````diff
@@ -0,0 +1,48 @@
+After that, follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):
+1. If there are other resources (e.g. `Certificate`s created by ingress-shim, `CertificateRequest`s, etc.) that aren't stored in Git and may have been created using the deprecated APIs, run
+
+   \```bash
+   kubectl get <resource_name> -oyaml | kubectl apply -f -
+   \```
````
suggestion: I don't think we need the backslashes here; I only added them in my suggestion so the github comment would format correctly.

SgtCoDFish reviewed (changes requested):
Just a couple more things; I should've been clearer before about what I meant about writing cert-manager, sorry!

(I'd make these changes myself and commit them to your branch but the required rebase makes that tricky to do - I understand why you've not rebased yet though!)

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,47 @@
+type: "docs"
+---
+
+We have deprecated the following cert-manager APIs: 
+- `cert-manager.io/v1alpha2`
+- `cert-manager.io/v1alpha3`
+- `cert-manager.io/v1beta1`
+- `acme.cert-manager.io/v1alpha2`
+- `acme.cert-manager.io/v1alpha3`
+- `acme.cert-manager.io/v1beta1`
+These APIs will be removed in cert-manager `v1.6.0`.
```
suggestion: tested locally: this doesn't render correctly as a list without spaces:

```suggestion
We have deprecated the following cert-manager APIs: 

- `cert-manager.io/v1alpha2`
- `cert-manager.io/v1alpha3`
- `cert-manager.io/v1beta1`
- `acme.cert-manager.io/v1alpha2`
- `acme.cert-manager.io/v1alpha3`
- `acme.cert-manager.io/v1beta1`

These APIs will be removed in cert-manager `v1.6.0`.
```

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,47 @@
+
+1. Familiarize yourself with the [official Kubernetes documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects) on CRD versioning.
+1. Make sure your installed cert-manager deployment is `v1.0.0` or later.
+2. Make sure any version-controlled cert-manager custom resource config files that refer to the deprecated APIs are updated to use the `cert-manager.io/v1` API and re-applied. This should update stored versions of the given resources.
```
question: Should this also reference `acme.cert-manager.io/v1`, since we also mention deprecating `acme.cert-manager.io/v1alpha2` etc in the list above?

irbekrm replied:
I didn't add `acme.cert-manager.io/v1` to keep this shorter- that API only has `Order`s and `Challenge`s and I couldn't think of any scenario where those would be in Git. Happy to be convinced otherwise :)

SgtCoDFish replied:
I think your judgement is sound there, just wanted to ask the question. Happy with it as-is :+1:

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,47 @@
+1. Familiarize yourself with the [official Kubernetes documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#writing-reading-and-updating-versioned-customresourcedefinition-objects) on CRD versioning.
+1. Make sure your installed cert-manager deployment is `v1.0.0` or later.
+2. Make sure any version-controlled cert-manager custom resource config files that refer to the deprecated APIs are updated to use the `cert-manager.io/v1` API and re-applied. This should update stored versions of the given resources.
+After that, follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):
```
suggestion: tested locally: this doesn't render correctly without a space above it:

```suggestion

After that, follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):
 ```

SgtCoDFish replied:
![render](https://user-images.githubusercontent.com/1972547/120334477-9270cf80-c2e8-11eb-9f45-73b263349097.png)

SgtCoDFish commented on content/en/docs/installation/upgrading/remove-deprecated-apis.md:
```diff
@@ -0,0 +1,47 @@
+2. Make sure any version-controlled cert-manager custom resource config files that refer to the deprecated APIs are updated to use the `cert-manager.io/v1` API and re-applied. This should update stored versions of the given resources.
+After that, follow the official Kubernetes documentation [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version):
+1. If there are other resources (e.g. `Certificate`s created by ingress-shim, `CertificateRequest`s, etc.) that aren't stored in Git and may have been created using the deprecated APIs, run
+
```
suggestion: tested locally: this blank line needs to be removed for this to render correctly

```suggestion
```

irbekrm replied:
Thanks for spotting all these issues 🙏🏼  🙏🏼 !

SgtCoDFish reviewed (changes requested):
aaaand a few bits I've noticed from running the code locally to see how it looks

SgtCoDFish commented:
This looks great now, but I guess I can't really LGTM it with the rebase still needing to be done? Happy to do that once it's ready to merge, though :+1:

irbekrm commented:
> I guess I can't really LGTM it with the rebase still needing to be done?

Thanks a lot for the review! I have rebased the release-next branch off the latest master. I realized that this PR still depends on #570 getting merged first as that adds the release notes file for `v1.4`

wallrj commented:
Drive by comments:
* How come this PR has so many unrelated changes? Looks like release-next is already almost equal with master.
* Maybe add a note to this page about the deprecation of api versions and link to the process for upgrading: https://cert-manager.io/docs/contributing/crds/#versions

irbekrm commented:
> How come this PR has so many unrelated changes? Looks like release-next is already almost equal with master.
Maybe add a note to this page about the deprecation of api versions and link to the process for upgrading: https://cert-manager.io/docs/contributing/crds/#versions

Thanks @wallrj ! I wasn't even aware of the https://cert-manager.io/docs/contributing/crds/#versions file.

I have updated the file and rebased the PR- this now has the correct number of commits and could be merged 😅

SgtCoDFish commented on content/en/docs/contributing/crds.md:
```diff
@@ -20,11 +20,10 @@ This will also update the version conversion code if needed.
 ## Versions
 
-cert-manager at time of writing has 4 CRD versions in use.
+cert-manager at time of writing has 4 CRD versions in use- `v1alpha2`, `v1alpha3`, `v1beta1` and `v1`. `v1alpha2`, `v1alpha3` and `v1beta1` are getting deprecated in cert-manager `v1.4.0` and will be removed in `v1.6.0`.
```
suggestion: would this look better as a list?

```suggestion
cert-manager currently has 4 CRD versions in use:

- `v1`
- `v1beta1` (deprecated in cert-manager `v1.4.0`, removed `v1.6.0`)
- `v1alpha3` (deprecated in cert-manager `v1.4.0`, removed `v1.6.0`)
- `v1alpha2` (deprecated in cert-manager `v1.4.0`, removed `v1.6.0`)
```

irbekrm replied:
Thanks a lot @SgtCoDFish 🙏🏼

SgtCoDFish commented on content/en/docs/contributing/crds.md:
```diff
@@ -34,6 +33,8 @@ We also have an internal API version, it lives at [`//pkg/internal/apis`](https:
 This is a version that is only used for validation and conversion, controllers should not use it as it is not meant to be user-friendly and not stable.
 However all new fields also have to be added here for the conversion logic to work.
 
+See the [official Kubernetes docs for CRD versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/) to understand conversion, which version are stored and served etc.
```
praise: love this addition, super useful!

SgtCoDFish reviewed (commented):
minor formatting suggestion, non-blocking!

SgtCoDFish reviewed (approved):
/lgtm

irbekrm commented:
Note to self: there are also some field changes between the API versions. But I think that would already be covered by the bit that asks users to update config files in Git.

irbekrm commented:
/cherry-pick-approve

irbekrm commented:
/cherry-pick-approved

irbekrm commented:
woooh got merged 🥳
